kani-compiler/src/kani_middle/coercion.rs

```diff
@@ -0,0 +1,188 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! This module contains methods that help us process coercions.
+//! There are many types of coercions in rust, they are described in the
+//! [RFC 401 Coercions](https://rust-lang.github.io/rfcs/0401-coercions.html).
+//!
+//! The more complicated coercions are DST Coercions (aka Unsized Coercions). These coercions
+//! allow rust to create references to dynamically sized types (such as Traits and Slices) by
+//! casting concrete sized types. Unsized coercions can also be used to cast unsized to unsized
+//! types. These casts work not only on the top of references, but it also handle
+//! references inside structures, allowing the unsized coercions of smart pointers. This last
+//! piece was defined by
+//! [RFC 982 DST Coercion](https://rust-lang.github.io/rfcs/0982-dst-coercion.html).
+
+use rustc_hir::lang_items::LangItem;
+use rustc_middle::traits::{ImplSource, ImplSourceUserDefinedData};
+use rustc_middle::ty::adjustment::CustomCoerceUnsized;
+use rustc_middle::ty::TypeAndMut;
+use rustc_middle::ty::{self, ParamEnv, TraitRef, Ty, TyCtxt};
+use rustc_span::symbol::Symbol;
+use tracing::trace;
+
+/// Extract the pair (`T`, `U`) for a unsized coercion where type `T` implements `Unsize<U>`.
+/// I.e., `U` is either a trait or a slice.
+/// For more details, please refer to:
+/// <https://doc.rust-lang.org/reference/type-coercions.html#unsized-coercions>
+///
+/// This is used to determine the vtable implementation that must be tracked by the fat pointer.
+///
+/// For example, if `&u8` is being converted to `&dyn Debug`, this method would return:
+/// `(u8, dyn Debug)`.
+///
+/// There are a few interesting cases (references / pointers are handled the same way):
+/// 1. Coercion between `&T` to `&U`.
+///    - This is the base case.
+///    - In this case, we extract the type that is pointed to.
+/// 2. Coercion between smart pointers like `Rc<T>` to `Rc<U>`.
+///    - Smart pointers implement the `CoerceUnsize` trait.
+///    - Use CustomCoerceUnsized information to traverse the smart pointer structure and find the
+///      underlying pointer.
+///    - Use base case to extract `T` and `U`.
+/// 3. Coercion between `&Wrapper<T>` to `&Wrapper<U>`.
```

Comment:
`Wrapper` here is a standin for some user defined type I assume? Maybe give an example of a user defined DST where this happens